// firebase-tools/lib/init/features/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apphosting = exports.hostingGithub = exports.remoteconfig = exports.project = exports.extensions = exports.emulators = exports.storage = exports.hosting = exports.functions = exports.firestore = exports.database = exports.account = void 0;
var account_1 = require("./account");
Object.defineProperty(exports, "account", { enumerable: true, get: function () { return account_1.doSetup; } });
var database_1 = require("./database");
Object.defineProperty(exports, "database", { enumerable: true, get: function () { return database_1.doSetup; } });
var firestore_1 = require("./firestore");
Object.defineProperty(exports, "firestore", { enumerable: true, get: function () { return firestore_1.doSetup; } });
var functions_1 = require("./functions");
Object.defineProperty(exports, "functions", { enumerable: true, get: function () { return functions_1.doSetup; } });
var hosting_1 = require("./hosting");
Object.defineProperty(exports, "hosting", { enumerable: true, get: function () { return hosting_1.doSetup; } });
var storage_1 = require("./storage");
Object.defineProperty(exports, "storage", { enumerable: true, get: function () { return storage_1.doSetup; } });
var emulators_1 = require("./emulators");
Object.defineProperty(exports, "emulators", { enumerable: true, get: function () { return emulators_1.doSetup; } });
var extensions_1 = require("./extensions");
Object.defineProperty(exports, "extensions", { enumerable: true, get: function () { return extensions_1.doSetup; } });
var project_1 = require("./project");
Object.defineProperty(exports, "project", { enumerable: true, get: function () { return project_1.doSetup; } });
var remoteconfig_1 = require("./remoteconfig");
Object.defineProperty(exports, "remoteconfig", { enumerable: true, get: function () { return remoteconfig_1.doSetup; } });
var github_1 = require("./hosting/github");
Object.defineProperty(exports, "hostingGithub", { enumerable: true, get: function () { return github_1.initGitHub; } });
var apphosting_1 = require("./apphosting");
Object.defineProperty(exports, "apphosting", { enumerable: true, get: function () { return apphosting_1.doSetup; } });
